data/tutorials/language/3ds_05_seq.md

```diff
@@ -325,124 +322,201 @@ This is the second line.
+partially applied function as an argument. This allows for lazy evaluation,
+ensuring that elements are fetched one at a time instead of forcing the entire
+sequence to be evaluated upfront.
+
+#### Example: `Seq.iter` (Consumer)
+
+```ocaml
+# let print_seq s = Seq.iter print_int s;;
+val print_seq : int Seq.t -> unit = <fun>
 
 ```
-let primes =
-  Seq.unfold_skippable
-    begin fun (prime_list, n) ->
-      if List.for_all (fun p -> n mod p <> 0) prime_list then
-        Yield (n, (n::prime_list, n+1))
-      else
-        Skip (prime_list, n+1)
-    end
-    ([], 2)
+
+In `print_seq`, `Seq.iter` takes a function (`print_int`) and applies it to each
+element as they are generated. If `Seq.iter` were to be fully applied, it would
+start consuming elements immediately. The consumer function (`print_int`) is a
+partially applied function that only takes one argument at a time.
+
+### Sequence Producers: Partially Applied Functions as Results
+
+A producer is a function that **generates** a sequence. Producers return a
+partially applied function so that elements are only computed when needed.  This
+ensures lazy evaluation and avoids unnecessary computation.
+
+#### Example: `Seq.unfold` (Producer)
+
+```ocaml
+ # let naturals =
+  Seq.unfold (fun x -> Some (x, x + 1)) 0;;
+val naturals : int Seq.t = <fun>
 ```
--->
 
-## Sequences Are Functions
+`Seq.unfold` returns a function of type `unit -> 'a Seq.node`, making it a lazy
+producer. Each time this function is called, a new element is produced.
+
+To recap, consumers fetch elements only when needed and producers delay
+computation until elements are requested. This pattern ensures that sequences
+remain lazy.
+
+## Be Aware of Seq.Cons vs Seq.cons
+
+The `Seq` module in the OCaml Standard Library contains two version of
+"cons-ing" that warrant special attention due to their similar names and distinct
+behavior.
+
+We have already been introduced to the `Seq.Cons` variant constructor. As a
+refresher, below is the type declaration for the `Seq` module:
+
+ ``` ocaml
+# type +'a node =
+    | Nil
+    | Cons of 'a * 'a t
+  and 'a t = unit -> 'a node;;
+type 'a node = Nil | Cons of 'a * 'a t
+and 'a t = unit -> 'a node
+
+ ```
 
-The `Seq` module contains this definition:
+The other version of "cons-ing" supported by the Standard Library is the
+function `Seq.cons` (with a lowercase "c") with the following value declaration:
 
 ```ocaml
 val cons : 'a -> 'a Seq.t -> 'a Seq.t
 ```
 
-Although `Seq.cons x seq` and `Seq.Cons (x, seq)` are the same, `Seq.cons` is a function and `Seq.Cons` is a variant's constructor, which is not the same in OCaml. This can lead to subtle bugs. This section illustrates this.
-<!--
-No need to introduce another mathematical sequence, we can reuse earlier examples (better for pedagogy):
+From the signature, we gather that it is a function that takes two parameters, a
+value and a sequence. Its function declaration is the following:
```

Comment:
```suggestion
value and a sequence. Its definition is the following:
```

Beware not to confuse “declaration” and “definition”
* a ”declaration” says “this is a name”
* a “definition” says “this is what this names stands for”

Definitions are declarations, but not the other way around